src/sidebar/components/hypothesis-app.js

```diff
@@ -112,6 +112,14 @@ function HypothesisAppController(
    *   completes. For non-OAuth logins, always resolves immediately.
    */
   this.login = function() {
+    // Clear the groups so that the groups service treats this load
+    // as the "firstLoad" and overrides the focused group. This is
+    // necessary for direct-linked private groups since this group
+    // won't be focused until the user logs in (because it doesn't
+    // exist in the list when the user is logged out).
+    if (store.getState().directLinkedGroupFetchFailed) {
+      store.clearGroups();
+    }
```

Comment:
Here is one way that the logic might be kept within the groups service:

```diff
diff --git a/src/sidebar/components/hypothesis-app.js b/src/sidebar/components/hypothesis-app.js
index 2fe930b6..1467be32 100644
--- a/src/sidebar/components/hypothesis-app.js
+++ b/src/sidebar/components/hypothesis-app.js
@@ -112,14 +112,6 @@ function HypothesisAppController(
    *   completes. For non-OAuth logins, always resolves immediately.
    */
   this.login = function() {
-    // Clear the groups so that the groups service treats this load
-    // as the "firstLoad" and overrides the focused group. This is
-    // necessary for direct-linked private groups since this group
-    // won't be focused until the user logs in (because it doesn't
-    // exist in the list when the user is logged out).
-    if (store.getState().directLinkedGroupFetchFailed) {
-      store.clearGroups();
-    }
     if (serviceConfig(settings)) {
       // Let the host page handle the login request
       bridge.call(bridgeEvents.LOGIN_REQUESTED);
diff --git a/src/sidebar/services/groups.js b/src/sidebar/services/groups.js
index 2affc64e..d57f6741 100644
--- a/src/sidebar/services/groups.js
+++ b/src/sidebar/services/groups.js
@@ -174,6 +174,14 @@ function groups(
       uri = getDocumentUriForGroupSearch();
     }
     const directLinkedGroup = settings.group;
+
+    // Use the direct link if this is the first load, or if a previous attempt
+    // to fetch a direct-linked group failed. The latter can happen if the
+    // direct link is to a private group and the user is logged out.
+    const useDirectLink =
+      store.allGroups().length === 0 ||
+      store.getState().directLinkedGroupFetchFailed;
+
     return uri
       .then(uri => {
         const params = {
@@ -247,7 +255,7 @@ function groups(
         const prevFocusedGroup = localStorage.getItem(STORAGE_KEY);
 
         store.loadGroups(groups);
-        if (isFirstLoad && groups.some(g => g.id === directLinkedGroup)) {
+        if (useDirectLink && groups.some(g => g.id === directLinkedGroup)) {
           store.focusGroup(directLinkedGroup);
         } else if (isFirstLoad && groups.some(g => g.id === prevFocusedGroup)) {
           store.focusGroup(prevFocusedGroup);
```

Are there any scenarios where this would result in the wrong thing happening?

Reply:
See below, we still need to handle the race condition, otherwise this doesn't actually load the groups for the private group but rather the group that was focused on when the user was logged out.